pkg/fx/config/provider.go

```diff
@@ -0,0 +1,291 @@
+	}
+
+	// Parse indexing service proofs if provided
+	var indexingProofs delegation.Proofs
+	if cfg.IndexingServiceProof != "" {
+		dlg, err := delegation.Parse(cfg.IndexingServiceProof)
+		if err != nil {
+			return app.AppConfig{}, fmt.Errorf("parsing indexing service proof: %w", err)
+		}
+		indexingProofs = delegation.Proofs{delegation.FromDelegation(dlg)}
+	}
+
+	externalConfig.IndexingService = app.IndexingServiceConfig{
+		Connection: indexingConn,
+		Proofs:     indexingProofs,
+	}
+
+	// PDP server (optional)
+	if cfg.PDPServerURL != "" {
+		pdpURL, err := url.Parse(cfg.PDPServerURL)
+		if err != nil {
+			return app.AppConfig{}, fmt.Errorf("parsing PDP server URL: %w", err)
+		}
+		externalConfig.PDPServer = &app.PDPServerConfig{
+			URL:      pdpURL,
+			ProofSet: cfg.ProofSet,
+		}
+	}
+
+	// Create services config
+	servicesConfig := app.ServicesConfig{
+		Publisher: app.PublisherConfig{
+			// Will be populated below
+		},
+		ServicePrincipalMapping: cfg.ServicePrincipalMapping,
+	}
+
+	// Build publisher config
+	pubMaddr, err := maurl.FromURL(publicURL)
+	if err != nil {
+		return app.AppConfig{}, fmt.Errorf("converting public URL to multiaddr: %w", err)
+	}
+
+	// Parse IPNI announce URLs
+	var announceURLs []url.URL
+	for _, s := range cfg.IPNIAnnounceURLs {
+		u, err := url.Parse(s)
+		if err != nil {
+			return app.AppConfig{}, fmt.Errorf("parsing IPNI announce URL %s: %w", s, err)
+		}
+		announceURLs = append(announceURLs, *u)
+	}
+
+	// Handle blob address for PDP
+	var blobAddr multiaddr.Multiaddr
+	if cfg.PDPServerURL != "" {
+		pdpURL, err := url.Parse(cfg.PDPServerURL)
+		if err != nil {
+			return app.AppConfig{}, fmt.Errorf("parsing PDP server URL: %w", err)
+		}
+		curioAddr, err := maurl.FromURL(pdpURL)
+		if err != nil {
+			return app.AppConfig{}, fmt.Errorf("converting PDP URL to multiaddr: %w", err)
+		}
+		pieceAddr, err := multiaddr.NewMultiaddr("/http-path/" + url.PathEscape("piece/{blobCID}"))
+		if err != nil {
+			return app.AppConfig{}, fmt.Errorf("creating piece multiaddr: %w", err)
+		}
+		blobAddr = multiaddr.Join(curioAddr, pieceAddr)
+	}
+
+	servicesConfig.Publisher = app.PublisherConfig{
+		PublicMaddr:   pubMaddr,
+		AnnounceMaddr: pubMaddr,
+		BlobMaddr:     blobAddr,
+		AnnounceURLs:  announceURLs,
+	}
+
+	identity, err := ReadPrivateKeyFromPEM(cfg.KeyFile)
+	if err != nil {
+		return app.AppConfig{}, fmt.Errorf("failed to read private key from PEM file: %w", err)
+	}
+
+	return app.AppConfig{
+		Identity: app.IdentityConfig{
+			Signer: identity,
+		},
+		Server:   serverConfig,
+		Storage:  storageConfig,
+		External: externalConfig,
+		Services: servicesConfig,
+	}, nil
+}
+
+// ProvideUploadServiceConnection provides the upload service connection for backward compatibility
+func ProvideUploadServiceConnection(cfg app.AppConfig) client.Connection {
+	return cfg.External.UploadService.Connection
+}
+
+func ReadPrivateKeyFromPEM(path string) (principal.Signer, error) {
```

Comment:
there is a duplicate of this function in the `cliutil` package that other packages use. Is that because these changes are an intermediate step and one of them will be removed once everything is consolidated?

Please remove it if that is not the case. If it needs to be kept, make it private, as it is only used within the package.